fuel-block-producer/src/block_producer/transaction_selector.rs

```diff
@@ -1,41 +1,164 @@
     //
     // Future improvements to this algorithm may take into account the parallel nature of
     // transactions to maximize throughput.
-    let mut used_block_space = Word::MIN;
+    let mut used_block_space: Word = 0;
 
-    // ensure txs are sorted by a fee
-    includable_txs.sort_by(|a, b| a.max_fee().cmp(&b.max_fee()));
+    // Sort transactions by gas price, highest first
+    includable_txs.sort_by_key(|a| Reverse(a.transaction().gas_price()));
 
+    // Pick as many transactions as we can fit into the block (greedy)
     includable_txs
         .into_iter()
-        .take_while(|tx| {
+        .filter(|tx| {
             let tx_block_space = tx.max_fee();
-            let new_used_space = used_block_space.saturating_add(tx_block_space);
-            let hit_end = new_used_space > config.max_gas_per_block;
-            used_block_space = new_used_space;
-            !hit_end
+            if let Some(new_used_space) = used_block_space.checked_add(tx_block_space) {
+                if new_used_space <= config.max_gas_per_block {
+                    used_block_space = new_used_space;
+                    true
+                } else {
+                    false
+                }
+            } else {
+                false
+            }
         })
         .collect()
 }
 
 #[cfg(test)]
 mod tests {
+    use fuel_core_interfaces::common::{
+        fuel_asm::Opcode,
+        fuel_crypto::rand::{thread_rng, Rng},
+        fuel_tx::{ConsensusParameters, Output, TransactionBuilder},
+        fuel_vm::consts::REG_ONE,
+    };
+    use itertools::Itertools;
+
     use super::*;
 
+    #[derive(Debug, Clone, Copy, PartialEq)]
+    struct TxGas {
+        pub price: u64,
+        pub limit: u64,
+    }
+
+    /// A test helper that generates set of txs with given gas prices and limits and runs
+    /// `select_transactions` against that, returning the list of selected gas price, limit pairs
+    fn make_txs_and_select(txs: &[TxGas], block_gas_limit: Word) -> Vec<TxGas> {
+        let mut rng = thread_rng();
+
+        let txs = txs
+            .iter()
+            .map(|tx_gas| {
+                CheckedTransaction::check(
+                    TransactionBuilder::script(
+                        vec![Opcode::RET(REG_ONE)].into_iter().collect(),
+                        vec![],
+                    )
+                    .gas_price(tx_gas.price)
+                    .gas_limit(tx_gas.limit)
+                    .add_unsigned_coin_input(
+                        rng.gen(),
+                        rng.gen(),
+                        1_000_000,
+                        Default::default(),
+                        Default::default(),
+                        0,
+                    )
+                    .add_output(Output::Change {
+                        to: Default::default(),
+                        amount: 0,
+                        asset_id: Default::default(),
+                    })
+                    .finalize(),
```

Comment:
use finalized_checked() to simplify this conversion: https://github.com/FuelLabs/fuel-tx/blob/master/src/builder.rs#L211